Add `CurvedScalarWave` executable that observes a scalar field in Kerr-Schild background

nikwit opened the pull request:
This adds an executable that observes the tails of a scalar wave on a Kerr-Schild background similar to the simulations of https://arxiv.org/abs/gr-qc/0305027.

It adds an observation of a surface integral of Psi^2 over a  spherical shell to the executable.

nikwit committed:
Add missing include

nikwit committed:
Add scalar wave on Kerr-Schild background executable

nilsdeppe commented on src/Evolution/Executables/CurvedScalarWave/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@
 # See LICENSE.txt for details.
 
 set(LIBS_TO_LINK
+  ApparentHorizons #necessary for interpolation onto surface
```
Can you add these link dependencies only to the executable that needs them?

nilsdeppe commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -110,6 +120,21 @@ class er;
 /// \endcond
 
+// We want to integrate psi^2 on a spherical shell
+struct PsiSquared : db::SimpleTag {
```
Would it be useful to stick this into `Evolution/Systems/CurvedScalarWave/PsiSquared.hpp`? Then it's easily tested too

nilsdeppe commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -142,6 +167,35 @@ struct EvolutionMetavars {
       typename system::gradient_variables>;
 
+  static constexpr bool interpolate = volume_dim == 3;
+  // We want to observe on a spherical shell of radius r so we use a Kerr
```
@markscheel do we not have a target for just a spherical surface? If not, should we add one? It seems a bit confusing to use a Kerr horizon just to get a sphere

nilsdeppe commented on tests/InputFiles/CurvedScalarWave/ScalarWaveKerrSchild.yaml:
```diff
@@ -0,0 +1,100 @@
+# Distributed under the MIT License.
```
I don't love the idea of adding another input file test (right now) because it means it's another executable we need to compile during CI. I think once we have initial data being controlled from the input file (hopefully ~1 month time scale) then this won't be an issue anymore. Maybe we omit for now and add once the initial data change is done?

nikwit committed:
Add Sphere InterpolationTarget

nikwit committed:
Add PsiSquared

nikwit commented:
@nilsdeppe I addressed your suggestions, thanks for the review. 

Since the changes split the PR up into additional commits for the `Sphere` interpolation target and the `PsiSquared` tag, I did not add this as a fixup commit. I removed the input file for the Kerr-Schild background. The unit test for the `Sphere` interpolation target is taken from the Kerr-Schild horizon unit test and adjusted.

nilsdeppe commented on src/ParallelAlgorithms/Interpolation/Targets/Sphere.cpp:
```diff
@@ -0,0 +1,32 @@
+
+namespace intrp::OptionHolders {
+Sphere::Sphere(size_t l_max, std::array<double, 3> center, double radius,
+               const Options::Context& /*context*/)
```
you can just omit the `Options::Context` completely if you don't use it as far as I know

nilsdeppe commented on src/ParallelAlgorithms/Interpolation/Targets/Sphere.hpp:
```diff
@@ -0,0 +1,146 @@
+  using is_sequential = std::false_type;
+  using frame = Frame;
+
+  using simple_tags = typename StrahlkorperTags::items_tags<Frame>;
```
I think this is an issue with my understanding of the interpolation framework, but how does one end up controlling what simple and compute tags are interpolated? Naively I would've expected this to be something like `typename InterpolationTagTag::item_tags`. The `LineSegment` also doesn't seem to have these type aliases, which confuses me more.

nikwit replied:
These compute tags are added to the `DataBox` of `InterpolationTarget` during `intrp::Actions::InitializeInterpolationTarget`. They are then needed during the call to `points` (`StrahlkorperTags::ComputeCartesianCoords` and its argument tags and argument tags' argument tags, ...). 

The `LineSegment` computes `points` directly and therefore does not need anything in the `DataBox`.

This is just for calculating the points to which to interpolate. The quantities that are then interpolated onto the points need to be specified in the MetaVariables in `compute_items_on_target`.

nilsdeppe commented on src/Evolution/Systems/CurvedScalarWave/PsiSquared.hpp:
```diff
@@ -0,0 +1,29 @@
+#include "Utilities/Gsl.hpp"
+
+namespace CurvedScalarWave::Tags {
+struct PsiSquared : db::SimpleTag {
```
did you want to add a short doxygen comment just so these tags show up in the documentation?

nilsdeppe commented on src/Evolution/Executables/CurvedScalarWave/CMakeLists.txt:
```diff
@@ -24,6 +24,12 @@ set(LIBS_TO_LINK
+  ApparentHorizons
+  ParallelInterpolation
+)
+
```
[optional] only one extra line?

nilsdeppe reviewed (commented):
Looks good. Just some minor questions and suggestions. Feel free to squash anything you end up changing :)

nilsdeppe commented on src/Evolution/Systems/CurvedScalarWave/PsiSquared.hpp:
```diff
@@ -0,0 +1,35 @@
+#include "Utilities/Gsl.hpp"
+
+/*!
+ * \brief The square of the scalar field \f$\Psi\f$.
```
Looks like you accidentally put the dox comment on the namespace instead of the tag :)

nikwit replied:
oups, my bad

nilsdeppe commented on src/ParallelAlgorithms/Interpolation/Targets/Sphere.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "Options/Options.hpp"
+
+namespace intrp::OptionHolders {
+Sphere::Sphere(size_t l_max_in, std::array<double, 3> center_in,
```
you can `const` the arguments here. Only asking because the dox comment needs to be moved anyway

nilsdeppe reviewed (commented):
LGTM, couple small typo things you can squash immediately

nilsvu commented on src/ParallelAlgorithms/Interpolation/Targets/Sphere.hpp:
```diff
@@ -0,0 +1,146 @@
+    const size_t l_max = sphere.l_max;
+    // Make a spherical strahlkorper
+    ::Strahlkorper<Frame> strahlkorper(
+        l_max, l_max, DataVector{(l_max + 1) * (2 * l_max + 1), sphere.radius},
```
Can you replace the `DataVector` with just the `sphere.radius` double? `Strahkorper` has a constructor for that.

nilsvu commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -70,8 +71,18 @@
+#include "PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp"
 #include "PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp"
 #include "PointwiseFunctions/AnalyticData/Tags.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrHorizon.hpp"
```
you can delete this now I think

nilsvu commented:
Looked at the code, you can fix these in some other PR if you get the chance
